[GSoC] Function for I.L.D. for reaction loads added

Psycho-Pirate opened the pull request:
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
See #19069

#### Brief description of what is fixed or changed
Function `solve_for_ild_reactions()` is added which returns the reaction equations under the effect of a moving load.
Function `plot_ild_reactions()` is added which plots these equations.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.continuum_mechanics 
   * Added ability to calculate and plot I.L.D. Reaction Equations
<!-- END RELEASE NOTES -->

Psycho-Pirate committed:
function for ild reactions added

Psycho-Pirate commented:
Here's an example:
![ildrxm](https://user-images.githubusercontent.com/50932406/124662730-d433fd80-dec6-11eb-9477-d465965526f1.png)
```
>>> from sympy.physics.continuum_mechanics.beam import Beam
>>> from sympy import symbols
>>> E, I = symbols('E, I')
>>> R1, R2 = symbols('R1, R2')
>>> b = Beam(10, E, I)
>>> b.apply_load(R1, 0, -1)
>>> b.apply_load(R2, 10, -1)
>>> print(b.solve_for_ild_reactions(1,R1,R2))
{R1: 1 - X/10, R2: X/10}
>>> b.plot_ild_reactions()
```
![ildr1](https://user-images.githubusercontent.com/50932406/124663209-6d631400-dec7-11eb-9193-d3dd9683501f.png)
![ildr2](https://user-images.githubusercontent.com/50932406/124663214-6e944100-dec7-11eb-9dc6-a17312c64f70.png)

Psycho-Pirate commented:
@ishanaj @Sc0rpi0n101

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1561,6 +1562,62 @@ def plot_loading_results(self, subs=None):
+        l = self.length
+        C3 = Symbol('C3')
+        C4 = Symbol('C4')
+        X = symbols('X')
```
What is the purpose of this `X` when we have `self.variable`?

Psycho-Pirate replied:
In order to solve the slope and moment equations, a limit (lim x->L) is applied. Here 'x' represents the length of the beam. The uppercase 'X' represents the distance of the moving load from origin. I have also explained this in my gsoc proposal.
![Screenshot (6)](https://user-images.githubusercontent.com/50932406/125154115-08811580-e176-11eb-828f-9d8a13c4e216.png)

ishanaj replied:
Firstly the big `X` is something the user never defined and thus would never know what it is. So you might have to provide user information about this `X` or create a similar flow of how the small `x` is taken as a parameter when a Beam object is defined.

Secondly, according to what I understand, `x` and `X` both vary over the beam length, and represent any random cross-section over the length, so we can use the same variable for them, unless there is some issue in solving the equations.

Psycho-Pirate replied:
> Secondly, according to what I understand, `x` and `X` both vary over the beam length, and represent any random cross-section over the length, so we can use the same variable for them, unless there is some issue in solving the equations.

The only reason I added 'X' was to avoid confusion as X is not any distance on the beam, it is the distance at which the moving load is acting. But if it causes more confusion then it can be definitely removed.

Psycho-Pirate committed:
changes made

Psycho-Pirate commented:
In the recent commit I made the following changes:
* Added two different examples in both functions along with an image.
* The `plot_ild_reactions()` now raises an error if `solve_for_ild_reactions()` is not called
* All the reaction graphs are plotted on a single plot
* removed 'X' variable
* fixed sign convention
@Sc0rpi0n101 @ishanaj

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1561,6 +1562,129 @@ def plot_loading_results(self, subs=None):
+            raise KeyError("solve_for_ild_reactions() not called")
+
+        x = self.variable
+        p = []
```
Variable name should be something more relevant for code readability

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1561,6 +1562,129 @@ def plot_loading_results(self, subs=None):
+
+        x = self.variable
+        p = []
+        for i in self._ild_reactions:
```
Here too, it can be `for reaction, equation in self._ild_reactions` or something better than just `i`

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -131,6 +131,7 @@ def __init__(self, length, elastic_modulus, second_moment, area=Symbol('A'), var
         self._support_as_loads = []
         self._applied_loads = []
         self._reaction_loads = {}
+        self._ild_reactions = {}
```
A property can also be defined for the same for the user to access

ishanaj commented:
I am getting this plot for the example present in `plot_ild_reactions()`. The y-axis has -ve values even above the origin. Why is it so?

![image](https://user-images.githubusercontent.com/29159034/126030595-b9a38515-4bb5-4138-9d2e-7946cbb24284.png)

ishanaj commented:
Two cases that need to be considered: 
Case 1: If the user has already called `solve_for_reaction_loads()`, the current code seems to fail.

Case 2: If length is a symbol, or if there are any other symbols present, we might need to keep a `subs={}` parameter in the plotting part. I haven't tried Case 2, but I believe this condition will occur. You can test an example and share the results if it works.

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1561,6 +1562,129 @@ def plot_loading_results(self, subs=None):
+
+        """
+        if not self._ild_reactions:
+            raise KeyError("solve_for_ild_reactions() not called")
```
Why a keyerror? I think it should be a ValueError instead. Also, the message should be elaborated like: "I.L.D. reaction equations not found. Please use `solve_for_ild_reactions()` to generate the I.L.D. reaction equations"

Psycho-Pirate commented:
I have implemented most of the changes you suggested.


> I am getting this plot for the example present in `plot_ild_reactions()`. The y-axis has -ve values even above the origin. Why is it so?
> 
I think this depends on the plot function as I have tried plotting the same equation just by using the plot function and it still shows negative values above origin.


> Two cases that need to be considered:
> Case 1: If the user has already called `solve_for_reaction_loads()`, the current code seems to fail.
> 
> Case 2: If length is a symbol, or if there are any other symbols present, we might need to keep a `subs={}` parameter in the plotting part. I haven't tried Case 2, but I believe this condition will occur. You can test an example and share the results if it works.

There are some things that need to be discussed here, after that I will push all the changes together.

Psycho-Pirate committed:
changes made

Psycho-Pirate commented:
In the recent commit I have made the following changes:
* Added a property which returns ild_reactions
* Added an attribute `_original_load` which retains an unsubstituted copy of load function
* Added a helper function `_solve_for_ild_reactions()` which calculates and returns shear force and bending moment using original load
* Added subs functionality in the `plot_ild_reactions()` function
* Changed variable names to something relevant to the beam module

The `solve_for_ild_reactions` function is now independent of `solve_for_reaction_loads`

Here is a working example:
```
>>> from sympy import symbols
>>> from sympy.physics.continuum_mechanics.beam import Beam
>>> E, I = symbols('E, I')
>>> R_0, R_7, l = symbols('R_0, R_7, l')
>>> subs = {l:10, E:100, I:25}
>>> b = Beam(l, E, I)
>>> b.apply_support(0, 'roller')
>>> b.apply_support(7, 'roller')
>>> b.apply_load(5, 4, -1)

>>> b.solve_for_reaction_loads(R_0,R_7)

>>> b.solve_for_ild_reactions(1,R_0,R_7)
>>> b.plot_ild_reactions(subs)
```
![ildild](https://user-images.githubusercontent.com/50932406/126888900-e132ff1f-b87a-40ee-9088-f5e7cce88fb5.png)

Psycho-Pirate commented:
> I am getting this plot for the example present in `plot_ild_reactions()`. The y-axis has -ve values even above the origin. Why is it so?

I have opened an issue regarding this [here](https://github.com/sympy/sympy/issues/21777)

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1561,6 +1577,159 @@ def plot_loading_results(self, subs=None):
+
+        # Determining the equations and solving them.
+        self._ild_reactions = dict(zip(reactions, solution))
+        return self._ild_reactions
```
`solve_for_ild_reactions` probably shouldn't return anything, it should just update the value of `self._ild_reactions`.
It goes with the behavior of the other solve functions in the module.

ishanaj commented:
I have left a review. Apart from that you should also include some release notes with this PR.
Rest looks good to me!

Psycho-Pirate committed:
changes made

Psycho-Pirate commented:
> I have left a review. Apart from that you should also include some release notes with this PR.
> Rest looks good to me!

Done!

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1561,6 +1577,161 @@ def plot_loading_results(self, subs=None):
+
+        # Determining the equations and solving them.
+        self._ild_reactions = dict(zip(reactions, solution))
+        self._ild_reactions
```
This line is not required anymore then.

Psycho-Pirate committed:
line removed

Psycho-Pirate commented:
I have removed the line

ishanaj commented:
LGTM! Merging..!
